File: src/data/articles/examples/q05.py
import json
import os
import tempfile
from pathlib import Path


def commit(directory, state, fail_before_replace=False):
    target = Path(directory) / "context.json"
    with tempfile.NamedTemporaryFile(mode="w", dir=directory, delete=False) as file:
        pending = Path(file.name)
        json.dump(state, file)
        file.flush()
        os.fsync(file.fileno())
    try:
        if fail_before_replace:
            raise RuntimeError("模拟提交前中断")
        os.replace(pending, target)  # 同文件系统、单写者的原子替换。
    finally:
        pending.unlink(missing_ok=True)


with tempfile.TemporaryDirectory() as directory:
    commit(directory, {"version": 1, "summary": "old"})
    try:
        commit(directory, {"version": 2, "summary": "new"}, True)
    except RuntimeError:
        pass
    recovered = json.loads((Path(directory) / "context.json").read_text())
    assert recovered["version"] == 1
    print(json.dumps({"recovered_version": recovered["version"]}))

File: src/data/articles/examples/q05.ts
import assert from 'node:assert/strict';
import { mkdtempSync, openSync, writeFileSync, fsyncSync, closeSync, renameSync, rmSync, readFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';

function commit(directory: string, state: object, failBeforeReplace = false) {
  const pending = join(directory, 'context.pending');
  const fd = openSync(pending, 'w');
  try { writeFileSync(fd, JSON.stringify(state)); fsyncSync(fd); }
  finally { closeSync(fd); }
  try {
    if (failBeforeReplace) throw new Error('模拟提交前中断');
    renameSync(pending, join(directory, 'context.json')); // 同文件系统、单写者。
  } finally { rmSync(pending, { force: true }); }
}

const directory = mkdtempSync(join(tmpdir(), 'context-demo-'));
try {
  commit(directory, { version: 1, summary: 'old' });
  assert.throws(() => commit(directory, { version: 2, summary: 'new' }, true));
  const recovered = JSON.parse(readFileSync(join(directory, 'context.json'), 'utf8'));
  assert.equal(recovered.version, 1);
  console.log(JSON.stringify({ recovered_version: recovered.version }));
} finally { rmSync(directory, { recursive: true, force: true }); }
